### PR TITLE
Implementing extensions to the cudamatrix class

### PR DESCRIPTION
**Pull request by JensWehner**
This will allow a bit more Eigen like calculations, although not even close to the same extend. 

For `gemm` calls it allows to use either `block()` or `transpose()` on some of the `cudamatrices`

This is nowhere near a complete set of functionality, but it is a start. I would like to get this merged as a small step.

### CODE REVIEW

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@votca-bot changlog: added transpose to CudaMatrix

**Comment by JensWehner** [comment]
@votca-bot changelog: added transpose to CudaMatrix

File: include/votca/xtp/cudamatrix.h [hunk 1]
```diff
@@ -48,13 +48,34 @@ cudaError_t checkCuda(cudaError_t result);
 
 Index count_available_gpus();
 
+template <class M>
```
Comment (felipeZ) [inline include/votca/xtp/cudamatrix.h:51]: `CudaMatrixTranspose` only differs of `CudaMatrix` in the flag that identifies whether or not the matrix is transpose? I am sure that I am missing something behind the reason to create this new clas.

File: src/libxtp/openmp_cuda.cc [hunk 2]
```diff
@@ -197,8 +197,8 @@ void OpenMP_CUDA::A_TDA(const Eigen::MatrixXd& matrix,
     temp_[threadid].B->copy_to_gpu(matrix);
     cuda_pips_[threadid]->diag_gemm(*temp_[threadid].B, *temp_[threadid].A,
                                     *temp_[threadid].C);
-    cuda_pips_[threadid]->gemm(*temp_[threadid].B, *temp_[threadid].C,
```
Comment (felipeZ) [inline src/libxtp/openmp_cuda.cc:200]: The new syntax is easier to follow :+1:

File: include/votca/xtp/cudamatrix.h (hunk 1 above)
Comment (JensWehner) [reply include/votca/xtp/cudamatrix.h:51]: Yes, exactly. I wanted to reproduce the `Eigen` synatx without actually modifying the object, so calling `transpose()` creates a wrapper object, which shifts the transpose flag.

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@felipeZ could you have a look again

File: src/libxtp/cudapipeline.cc [hunk 3]
```diff
@@ -58,42 +58,19 @@ std::string cudaGetErrorEnum(cublasStatus_t error) {
...
-    transA = CUBLAS_OP_T;
-    k = int(A.rows());
-  }
-  cublasOperation_t transB = CUBLAS_OP_N;
-  int k2 = int(B.rows());
-  if (transpose_B) {
-    transB = CUBLAS_OP_T;
-    k2 = int(B.cols());
-  }
-
```
Comment (felipeZ) [inline src/libxtp/cudapipeline.cc:84]: Now I got the point of the `CudaMatrixTranspose`, great job!

File: include/votca/xtp/cudapipeline.h [hunk 4]
```diff
@@ -57,28 +57,94 @@ class CudaPipeline {
...
+  }
+  template <class M1, class M2>
+  void gemm(const M1 &A, const M2 &B, CudaMatrix &C, double beta) const {
+    gemm_internal(A, B, C, beta);
+  }
+
+  template <class M1, class M2>
+  void gemm(const M1 &A, const M2 &B, CudaMatrix &C) const {
+    gemm(A, B, C, 0.0);
+  }
```
Comment (felipeZ) [inline include/votca/xtp/cudapipeline.h:103]: nice!

**Review by felipeZ — APPROVED** [review]

**Pull request merged**